Optimize flat display lists

pcwalton opened the pull request:
Flat display lists were a 2x regression on the spheres demo. This patch series fixes that.

See the individual commits for more details.

r? @mrobinson

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9843)

<!-- Reviewable:end -->

pcwalton committed:
gfx: Box stacking contexts to minimize `memmove` traffic.

`memmove` was showing up high in the profile when concatenating and
shorting display lists. This change drastically reduces the `memmove`
cost in exchange for some minor additional allocation cost.

pcwalton committed:
gfx: Avoid copying stacking contexts around so much during stacking
context creation.

pcwalton committed:
gfx: Stop cloning clipping regions so much.

Clipping regions can contain vectors and so can be expensive to copy.

pcwalton committed:
gfx: Switch offsets to an FNV hasher.

SipHash traffic was showing up high in the profile.

Unfortunately, this necessitated a manual implementation of the serde
traits.

pcwalton committed:
layout: Minor whitespace cleanup.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify gfx and layout code, but no tests are modified. Please consider adding a test!

mrobinson commented on components/layout/display_list_builder.rs:
```diff
@@ -1644,18 +1638,24 @@ impl BlockFlowDisplayListBuilding for BlockFlow {
-            stacking_context.children.append(&mut floating);
-            contexts.push(stacking_context);
-            contexts.append(&mut positioned);
+            contexts[stacking_context_index].children = floating;
```
I wonder if there is some other way to do this without copying. I'm a little wary of this because it means that when you dump the display list, the output will not accurately reflect the structure of the stacking contexts. Is a contexts.append(...) here so expensive now that stacking contexts are boxed?

pcwalton replied:
I'm confused as to why the output will not reflect the structure of the stacking contexts—could you elaborate?

mrobinson replied:
Ah, sorry! I misunderstood this code.

pcwalton committed:
layout: Remove the `validate_display_list_geometry` debugging tool.

I don't think anyone was using it, and it's incompatible with taking
display lists out of flows.

pcwalton committed:
layout: Don't call `establishes_stacking_context()` so much.

That function is expensive because it has to check a lot.

pcwalton committed:
layout: Switch display list building from bottom-up to top-down.

This eliminates a lot of allocations and O(n^2) behavior.

mrobinson commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 486e4e0 has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson 486e4e03fcd23970b91f0c7f140312a735f4033b -->

bors-servo commented:
:hourglass: Testing commit 486e4e0 with merge c994acd...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/1905)

mrobinson commented:
Looks like a tidy failure:

```
./components/gfx/display_list/mod.rs:33: use statement is not in alphabetical order
	expected: serde::ser::{Serialize, Serializer}
	found: serde::ser::impls::MapIteratorVisitor
```

nox commented:
I also want a squash, and no commits like `layout: Minor whitespace cleanup.`, please.

Ms2ger commented:
Sorry, fat fingers. I don't see a reason to squash, though.

nox commented:
Yeah never mind that, thought that was a fix for the previous commit in same PR.

pcwalton commented:
@bors-servo: r=mrobinson

bors-servo commented:
:pushpin: Commit 117e92a has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson 117e92aefd0136e192edf132c9bd5c75bce9bd5f -->

bors-servo commented:
:hourglass: Testing commit 117e92a with merge 55fc48e...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1893), [gonk](http://build.servo.org/builders/gonk/builds/1936), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1912), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2192), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/611), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1965), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2018)
